### DISCUSSION TITLE
How does `Connection.prototype.useDb()` works under the hood?

### DISCUSSION THREAD

**Discussion by sadeghhosseini** [Automattic/mongoose#13266] [Q&A] [answered]
I am trying to understand the [Connection.prototype.useDb()](https://mongoosejs.com/docs/api/connection.html#Connection.prototype.useDb()) method better. I would appreciate it if someone could help me.

1. 
The mongoose document for [useDb()](https://mongoosejs.com/docs/api/connection.html#Connection.prototype.useDb()) states:
> Switches to a different database using the same connection pool.

Does this mean that if there is an available connection in the connection pool `useDb('db1')` will reuse the connection? in other words can a connection used by `useDb('db1')`, be reused by `useDb('db2')`? 
Or does the statement simply mean that `useDb('db1')` will create a new connection in the connection pool and can only reuse connections already assigned to `useDb('db1')`?

2. 
The mongoose document for [useDb()](https://mongoosejs.com/docs/api/connection.html#Connection.prototype.useDb()) states:
> [options.useCache=false] «Boolean» If true, cache results so calling useDb() multiple times with the same name only creates 1 connection object.

What happens in the following scenario(considering that [`maxIdleTimeMS`](https://www.mongodb.com/docs/drivers/node/current/faq/#how-can-i-prevent-a-slow-operation-from-delaying-other-operations-:~:text=Once%20a%20connection%20has%20been%20idle%20for%20maxIdleTimeMS%2C%20the%20connection%20pool%20removes%20and%20replaces%20it.%20This%20option%20defaults%20to%200%20(no%20limit).) defaults to 0 which means no limit)?
![image](https://user-images.githubusercontent.com/13155024/231394548-d4d7be03-a42d-490d-8af4-6ff248026d0c.png)

3. 
Considering:

> [options.useCache=false] «Boolean» If true, cache results so calling useDb() multiple times with the same name only creates 1 connection object.

and considering [this](https://www.mongodb.com/docs/drivers/node/current/faq/#how-can-i-prevent-a-slow-operation-from-delaying-other-operations-) from mongodb.com which states:
> MongoDB is synchronous and uses a single execution thread per socket, meaning that MongoDB will execute one single operation per socket at any point in time. Any other operation sent to that socket will have to wait until the current operation is finished. 

Does this mean 5 queries which are supposed to be executed on `mydb2`, using the connection from `conn.useDb('mydb2', { useCache: true })`, would have to wait in queue to be executed one after another? as opposed to  `conn.useDb('mydb2', { useCache: false })`, which means they will be executed on separate connections and in parallel?

**Comment by sadeghhosseini** [comment, marked as answer]
I did a little experiment and I found [these results](https://stackoverflow.com/a/76024044/2605781)
 
summary:
1. It seems like if there is an available connection(socket) in the connection pool conn.useDb will reuse the connection, in other words a connection used/created by conn.useDb('db1'), can be reused by conn.useDb('db2').
2. [options.useCache=true] works like this:
```
if the cached connection exists in the connection pool(it's not checked out of the pool(it's available and not currently being used))
  then the cached connection will get returned
else
  if connection pool is not empty
    a connection from the connection pool is returned
  else if maxPoolSize has not reached
     creates a new connection in the connection pool
  else 
    wait for a connection to get availabe
```

**Comment by vkarpov15** [comment]
1. Yes, if there's an available socket on the original connection, then `useDb('db2')` can use that socket. That's the purpose of `useDb()`: multiple connections can share the same sockets.
2. I think you're misunderstanding the purpose of `useCache`. `useCache` means that `conn.useDb('mydb2', { useCache: true }) === conn.useDb('mydb2', { useCache: true })`. `useDb()` still uses the same connection pool under the hood, regardless of `useCache`. `useCache` is just a convenience to avoid creating multiple connections to the same database using `useDb()`.
